dnf/transaction_sr.py

```diff
@@ -0,0 +1,359 @@
+
+            groups.append(group_data)
+
+        elif tsi.is_environment():
+            env = tsi.get_environment()
+
+            env_data = {
+                "action": tsi.action_name,
+                "id": env.getEnvironmentId(),
+            }
+
+            if tsi.action == libdnf.transaction.TransactionItemAction_INSTALL:
+                env_data["package_types"] = \
+                    libdnf.transaction.compsPackageTypeToString(env.getPackageTypes())
+
+            environments.append(env_data)
+
+    if rpms:
+        data["rpms"] = rpms
+
+    if groups:
+        data["groups"] = groups
+
+    if environments:
+        data["environments"] = environments
+
+    with open(filename, "w") as f:
+        json.dump(data, f, indent=4, sort_keys=True)
+        f.write("\n")
+
+    print(_("Transaction saved to {}.").format(filename))
+
+
+class TransactionReplay(object):
+    def __init__(self, base, fn):
+        self._base = base
+        self._filename = fn
+
+    def _verify_toplevel_json(self, replay_data):
+        fn = self._filename
+
+        if "version" not in replay_data:
+            raise TransactionFileError(fn, _('Missing key "{key}".'.format(key="version")))
+
+        if not isinstance(replay_data["version"], str):
+            raise TransactionFileError(fn, _('Unexpected type of "version", string expected.'))
+
+        check_version(replay_data["version"], fn)
+
+    def _replay_pkg_action(self, pkg_data):
+        try:
+            action = pkg_data["action"]
+            nevra = pkg_data["nevra"]
+            reason = pkg_data["reason"]
+        except KeyError as e:
+            raise TransactionError(
+                _('Missing object key "{key}" in an rpm.').format(key=e.args[0])
+            )
+
+        solution = dnf.subject.Subject(nevra).get_best_solution(
+            self._base.sack,
+            with_nevra=True,
+            with_provides=False,
+            with_filenames=False,
+            forms=[hawkey.FORM_NEVRA]
+        )
+
+        if solution["nevra"] is None:
+            raise TransactionError(_('Cannot find rpm nevra "{nevra}".').format(nevra=nevra))
+
+        query = solution["query"]
+        if not query:
+            raise TransactionError(_('Rpm nevra "{nevra}" was not found.').format(nevra=nevra))
+
+        # validate the reasons before running the transaction and cache them
+        try:
+            self._pkg_reason_cache[nevra] = \
+                libdnf.transaction.StringToTransactionItemReason(reason)
+        except IndexError as e:
+            raise TransactionError(
+                _('Unexpected value of package reason "{reason}" for rpm nevra "{nevra}".')
+                    .format(reason=reason, nevra=nevra)
+            )
+
+        if action in ("Install", "Upgrade", "Downgrade", "Reinstall"):
+            sltr = dnf.selector.Selector(self._base.sack).set(pkg=query.available())
+            self._base.goal.install(select=sltr, optional=False)
```

Comment:
Please could you add TODO: replace `optional=False` by `optional=not self._base.conf.strict`?

Reply:
Why a  TODO? We've already discussed this and I still believe it is correct to always be strict here. (I'm not repeating my reasoning here since I feel I've stated it multiple times already...)

Reply:
Because you told me, that you want to implement it later on, therefore TODO.
If you you need to refresh memory, here are basic points:
1. Subcommand `history replay` will be not usable on Fedora
2. Other history transaction commands are implemented by the same way
3. Other transaction commands (install, distrosync) are implemented by the same way